src/main/java/org/truffleruby/language/control/FrameOnStackNode.java

```diff
@@ -11,27 +11,25 @@
 
 import org.truffleruby.language.RubyContextSourceNode;
 import org.truffleruby.language.RubyNode;
-import org.truffleruby.language.locals.WriteFrameSlotNode;
-import org.truffleruby.language.locals.WriteFrameSlotNodeGen;
 
 import com.oracle.truffle.api.frame.FrameSlot;
 import com.oracle.truffle.api.frame.VirtualFrame;
 
 public class FrameOnStackNode extends RubyContextSourceNode {
 
     @Child private RubyNode child;
-    @Child private WriteFrameSlotNode writeMarker;
+    private final FrameSlot markerSlot;
 
     public FrameOnStackNode(RubyNode child, FrameSlot markerSlot) {
         this.child = child;
-        writeMarker = WriteFrameSlotNodeGen.create(markerSlot);
+        this.markerSlot = markerSlot;
     }
 
     @Override
     public Object execute(VirtualFrame frame) {
         final FrameOnStackMarker marker = new FrameOnStackMarker();
 
-        writeMarker.executeWrite(frame, marker);
+        frame.setObject(markerSlot, marker);
```

Comment:
There is one difference here, the node does `frame.getFrameDescriptor().setFrameSlotKind(frameSlot, FrameSlotKind.Object);`.
We should set the FrameSlot's kind in the translator (under `frameOnStackMarkerSlot = environment.declareVar(environment.allocateLocalTemp("frame_on_stack_marker"));`) to achieve the same.
I'm not sure how important it is, but it sounds better if the FrameSlot's kind is correctly set.

Reply:
Pushed as https://github.com/oracle/truffleruby/pull/2530/commits/6b0824ed36eb654ff26b50cced70349e57395bee